### ISSUE TITLE
missing symbols from embed.h when testing B::Generate

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6611]
Migrated from [rt.perl.org#22959](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22959) (status was 'resolved')

Searchable as RT22959$

**Comment by p5pRT** [comment]
### From jcromie@divsol.com

#### Created by jimc@harpo.jimc.earth



When doing 'make test' on B&#8203;::Generate on 5\.8\.1\-19893\,
I get missing symbols\, which are defined in embed\.h\,
and used successfully in core\.  For me at least\, theyre
not available to this XS module \(though other non\-CORE
XS's build ok\)\.

I dunno whether its a bug in CORE\, in B&#8203;::Generate\, or
a PEBCAK error\, but I thought it safest to flag it\,
in case it impacts 5\.8\.1\.


\[jimc@&#8203;harpo maintperl\]$ grep \-r fold\_constants \*
embed\.fnc&#8203;:p    |OP\*    |fold\_constants    |OP\* arg
embed\.h&#8203;:\#define fold\_constants        Perl\_fold\_constants
embed\.h&#8203;:\#define fold\_constants\(a\)    Perl\_fold\_constants\(aTHX\_ a\)
op\.c&#8203;:Perl\_fold\_constants\(pTHX\_ register OP \*o\)
op\.c&#8203;:    return fold\_constants\(o\);
op\.c&#8203;:    return fold\_constants\(\(OP \*\) unop\);
op\.c&#8203;:    return fold\_constants\(\(OP \*\)binop\);
op\.c&#8203;:    o = fold\_constants\(o\);
op\.c&#8203;:        return fold\_constants\(o\);
proto\.h&#8203;:PERL\_CALLCONV OP\*    Perl\_fold\_constants\(pTHX\_ OP\* arg\);

$ more \.patch
19893

The build below is on a minimally patched B\-Generate
\(to get it to compile\)

\[jimc@&#8203;harpo optimization\]$ diff \-ru B\-Generate\-1\.06 B\-Generate\-1\.06\-mod/
<details><summary>Inline Patch</summary>

```diff
diff -ru B-Generate-1.06/Build.PL B-Generate-1.06-mod/Build.PL
--- B-Generate-1.06/Build.PL	Tue Jun 18 21:14:58 2002
+++ B-Generate-1.06-mod/Build.PL	Fri Jun 20 18:57:39 2003
@@ -22,6 +22,7 @@
      module_name => 'B::Generate',
      license => 'unknown',
      dynamic_config => 0,
+     requires => { perl => '5.8.0' },
      );
 $build->create_build_script;
 
diff -ru B-Generate-1.06/lib/B/Generate.xs B-Generate-1.06-mod/lib/B/Generate.xs
--- B-Generate-1.06/lib/B/Generate.xs	Sun Jul 28 10:45:48 2002
+++ B-Generate-1.06-mod/lib/B/Generate.xs	Sun Jun 22 06:44:39 2003
@@ -460,7 +461,7 @@
 OP_find_cv(o)
 	B::OP	o
     CODE:
-	RETVAL = SvRV(find_cv_by_root((OP*)o));
+	RETVAL = (CV*) SvRV(find_cv_by_root((OP*)o));
     OUTPUT:
 	RETVAL
 
@@ -641,8 +642,8 @@
 
         o = CALL_FPTR(PL_check[type])(aTHX_ (OP*)o);
 
-        if (o->op_type == type)
-            o = fold_constants(o);
+	if (o->op_type == type)
+            o = (OP*) fold_constants(o);
 
     OUTPUT:
         o


```
</details>
[jimc@harpo B-Generate-1.06-mod]$ make
/usr/local/bin/perl5\.8\.1 \./Build
lib/B/Generate\.pm \-> blib/lib/B/Generate\.pm
cc \-c \-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-I/usr/include/gdbm \-g \-O2 \-I/usr/local/lib/perl5/5\.8\.1/i686\-linux\-thread\-multi/CORE \-o lib/B/Generate\.o lib/B/Generate\.c
lib/B/Generate\.bs \-> blib/arch/auto/B/Generate/Generate\.bs
cc \-shared \-L/usr/local/lib \-o blib/arch/auto/B/Generate/Generate\.so lib/B/Generate\.o
\[jimc@&#8203;harpo B\-Generate\-1\.06\-mod\]$ make test
/usr/local/bin/perl5\.8\.1 \./Build test
test\.\.\.\.Can't load 'blib/arch/auto/B/Generate/Generate\.so' for module B&#8203;::Generate&#8203;: blib/arch/auto/B/Generate/Generate\.so&#8203;: undefined symbol&#8203;: fold\_constants at /usr/local/lib/perl5/5\.8\.1/i686\-linux\-thread\-multi/DynaLoader\.pm line 229\.
 at test\.pl line 4

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.0 - Tue Jul  1 06:52:23 MDT 2003
It is being executed now by  Perl v5.8.1 - Mon Jun 30 12:06:36 MDT 2003.

Site configuration information for perl v5.8.1:

Configured by jimc at Mon Jun 30 12:06:36 MDT 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.20-18.7, archname=i686-linux-thread-multi
    uname='linux harpo.jimc.earth 2.4.20-18.7 #1 thu may 29 08:32:50 edt 2003 i686 unknown '
    config_args='-Doptimize=-g -O2 -Dusethreads -Uusrbininstall'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-g -O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.2 2.96-112.7.2)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil -lrt
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil -lrt
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.1:
    /usr/local/lib/perl5/5.8.1/i686-linux-thread-multi
    /usr/local/lib/perl5/5.8.1
    /usr/local/lib/perl5/site_perl/5.8.1/i686-linux-thread-multi
    /usr/local/lib/perl5/site_perl/5.8.1
    /usr/local/lib/perl5/site_perl/5.8.0/i686-linux-thread-multi
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.1:
    HOME=/home/jimc
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/bin:/usr/bin:/usr/bin/X11:/usr/local/bin:/usr/bin:/usr/java/j2sdk1.4.1_02/bin:/usr/local/mozilla:/usr/java/j2sdk1.4.0/bin:/home/jimc/bin:./bin:.:/usr/sbin:/sbin:/usr/X11R6/bin:/usr/java/j2sdk1.4.1_02/bin:/usr/java/j2sdk1.4.1_02/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

> \[jimc \- Sun Jul 13 22&#8203;:36&#8203;:22 2003\]&#8203;:
> 
> When doing 'make test' on B&#8203;::Generate on 5\.8\.1\-19893\,
> I get missing symbols\, which are defined in embed\.h\,
> and used successfully in core\.  For me at least\, theyre
> not available to this XS module \(though other non\-CORE
> XS's build ok\)\.

> I dunno whether its a bug in CORE\, in B&#8203;::Generate\, or
> a PEBCAK error\, but I thought it safest to flag it\,
> in case it impacts 5\.8\.1\.

> \[jimc@&#8203;harpo maintperl\]$ grep \-r fold\_constants \*
> embed\.fnc&#8203;:p     |OP\*    |fold\_constants |OP\* arg
> embed\.h&#8203;:\#define fold\_constants          Perl\_fold\_constants
> embed\.h&#8203;:\#define fold\_constants\(a\)       Perl\_fold\_constants\(aTHX\_ a\)
> op\.c&#8203;:Perl\_fold\_constants\(pTHX\_ register OP \*o\)
> op\.c&#8203;:    return fold\_constants\(o\);
> op\.c&#8203;:    return fold\_constants\(\(OP \*\) unop\);
> op\.c&#8203;:    return fold\_constants\(\(OP \*\)binop\);
> op\.c&#8203;:    o = fold\_constants\(o\);
> op\.c&#8203;:       return fold\_constants\(o\);
> proto\.h&#8203;:PERL\_CALLCONV OP\*       Perl\_fold\_constants\(pTHX\_ OP\* arg\);



> \[jimc@&#8203;harpo B\-Generate\-1\.06\-mod\]$ make
> /usr/local/bin/perl5\.8\.1 \./Build
> lib/B/Generate\.pm \-> blib/lib/B/Generate\.pm
> cc \-c \-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBUGGING \-fno\-
strict\-aliasing \-
> \\ I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-I/usr/include/
gdbm \-g \-O2
> \\ \-I/usr/local/lib/perl5/5\.8\.1/i686\-linux\-thread\-multi/CORE \-o lib/B/Generate\.o
> \\ lib/B/Generate\.c
> lib/B/Generate\.bs \-> blib/arch/auto/B/Generate/Generate\.bs
> cc \-shared \-L/usr/local/lib \-o blib/arch/auto/B/Generate/Generate\.so lib/B/Generate\.o
> \[jimc@&#8203;harpo B\-Generate\-1\.06\-mod\]$ make test
> /usr/local/bin/perl5\.8\.1 \./Build test
> test\.\.\.\.Can't load 'blib/arch/auto/B/Generate/Generate\.so' for module
> \\ B&#8203;::Generate&#8203;: blib/arch/auto/B/Generate/Generate\.so&#8203;: undefined symbol&#8203;: fold\_constants 
at
> \\ /usr/local/lib/perl5/5\.8\.1/i686\-linux\-thread\-multi/DynaLoader\.pm line 229\.
 at test\.pl line 4

The issue here is that C\<fold\_constants> is not part of the public API and is supposed to 
only be available only in the Perl core\.  A more complete look at embed\.h shows

&nbsp;    \#ifdef PERL\_CORE
&nbsp;    \#define find\_script             Perl\_find\_script
&nbsp;    \#define force\_list              Perl\_force\_list
&nbsp;    \#define fold\_constants          Perl\_fold\_constants
&nbsp;    \#endif

So\, given this\, I'd say this is a bug in B&#8203;::Generate\, rather than the Perl core\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From jcromie@divsol.com

Steve Peters via RT wrote&#8203;:

>>\[jimc \- Sun Jul 13 22&#8203;:36&#8203;:22 2003\]&#8203;:
>>
>>When doing 'make test' on B&#8203;::Generate on 5\.8\.1\-19893\,
>>I get missing symbols\, which are defined in embed\.h\,
>>and used successfully in core\.  For me at least\, theyre
>>not available to this XS module \(though other non\-CORE
>>XS's build ok\)\.
>>    
>>
>
>  
>
>The issue here is that C\<fold\_constants> is not part of the public API and is supposed to 
>only be available only in the Perl core\.  A more complete look at embed\.h shows
>
>    \#ifdef PERL\_CORE
>    \#define find\_script             Perl\_find\_script
>    \#define force\_list              Perl\_force\_list
>    \#define fold\_constants          Perl\_fold\_constants
>    \#endif
>
>So\, given this\, I'd say this is a bug in B&#8203;::Generate\, rather than the Perl core\.  
>
>  
>
The thing is\, it compiles \(with warnings\)\, and passes its tests \(with 
warnings\)
on 5\.8\.1\, and 5\.8\.0 \(IIRC\, not at hand\)

and 5\.8\.2 \( tested this with patched version to build and test clean\, 10 
tests\)
it does also for 5\.8\.\[34567\]

ths patch fixes the problem reported by doing this&#8203;:

\-        if \(o\->op\_type == type\)
\-            o = fold\_constants\(o\);
\+       if \(o\->op\_type == type\)
\+            o = \(OP\*\) Perl\_fold\_constants\(aTHX\_ o\);

in other words\, it does more or less what embed would do if it were part 
of the api\.


0\.01  Sat Dec 23 18&#8203;:00&#8203;:08 2000
&nbsp;        \- original version; created by h2xs 1\.20 with options
&nbsp;                \-A \-n B&#8203;::Generate

I want to note that this module predates 5\.6\.0\, and works \(passes tests\) 
on 5\.6\.2
&nbsp;     requires => \{ perl => '5\.5\.62' \}\,

Theres no embed\.fnc in 5\.6\.2\, but the old code 'fold\_constants' compiles 
properly there\.
Its not clear to me how and/or why fold\_constants didnt make the cut when
the API/non\-API choices were made\.
